lxc/console.go

```diff
@@ -201,3 +215,110 @@ func (c *cmdConsole) Console(d lxd.InstanceServer, name string) error {
 
 	return nil
 }
+
+func (c *cmdConsole) vga(d lxd.InstanceServer, name string) error {
+	// We currently use the control websocket just to abort in case of
+	// errors.
+	controlDone := make(chan struct{}, 1)
+	handler := func(control *websocket.Conn) {
+		<-controlDone
+		closeMsg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
+		control.WriteMessage(websocket.CloseMessage, closeMsg)
+	}
+
+	// Prepare the remote console
+	req := api.InstanceConsolePost{
+		Type: "vga",
+	}
+
+	consoleDisconnect := make(chan bool)
+	sendDisconnect := make(chan bool)
+	defer close(sendDisconnect)
+
+	consoleArgs := lxd.InstanceConsoleArgs{
+		Control:           handler,
+		ConsoleDisconnect: consoleDisconnect,
+	}
+
+	go func() {
+		<-sendDisconnect
+		close(consoleDisconnect)
+	}()
+
+	// Create a temporary unix socket mirroring the instance's spice
+	// socket.
+	path, err := ioutil.TempFile("", "*.spice")
+	if err != nil {
+		return err
+	}
+	err = os.Remove(path.Name())
+	if err != nil {
+		return err
+	}
+	path.Close()
+
+	socket := path.Name()
+
+	listener, err := net.Listen("unix", socket)
+	if err != nil {
+		return err
+	}
+	defer listener.Close()
+
+	op, connect, err := d.ConsoleInstanceDynamic(name, req, &consoleArgs)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				return
+			}
+			go func(conn io.ReadWriteCloser) {
+				err = connect(conn)
+				if err != nil {
+					select {
+					case controlDone <- struct{}{}:
+					}
+				}
+			}(conn)
+		}
+	}()
+
+	// Use either spicy or remote-viewer if available.
+	var cmd *exec.Cmd
+
+	spicy, err := exec.LookPath("spicy")
+	if err == nil {
+		cmd = exec.Command(spicy, fmt.Sprintf("--uri=spice+unix://%s", socket))
+	} else {
+		remoteViewer, err := exec.LookPath("remote-viewer")
+		if err == nil {
+			cmd = exec.Command(remoteViewer, fmt.Sprintf("spice+unix://%s", socket))
+		}
+	}
+
+	if cmd != nil {
+		cmd.Output()
+	} else {
+		fmt.Printf("\n"+i18n.G("No remote desktop client found, raw SPICE socket available at:")+"\n\n%s\n", socket)
+
+		ch := make(chan os.Signal, 10)
+		signal.Notify(ch, syscall.SIGINT)
+		<-ch
```

Comment:
This still isn't going to go well on Windows. I think we should just hang after showing the message. It would be consistent with what we're already doing in such fallback scenarios for web browsers and Candid.

Reply:
> This still isn't going to go well on Windows. I think we should just hang after showing the message. It would be consistent with what we're already doing in such fallback scenarios for web browsers and Candid.

How is "hanging" implemented there?

Reply:
Also, why do you think it's not going to go well on Windows? Perhaps using `signal.Notify(ch, os.Interrupt)` instead? Note that it'd be nice to catch a signal instead of exiting abruptly, so the graceful shutown at the end of the function can take place.

Reply:
Maybe just do something like `Hit enter to disconnect from the instance`?